lib/wallaroo/core/sink/connector_sink/connector_sink.pony

```diff
@@ -387,25 +722,94 @@ actor ConnectorSink is Sink
   ///////////////
   // CHECKPOINTS
   ///////////////
+
   fun ref checkpoint_state(checkpoint_id: CheckpointId) =>
     """
-    ConnectorSinks don't currently write out any data as part of the checkpoint.
+    Serialize hard state (i.e., can't afford to lose it) and send
+    it to the local event log and reset 2PC state.
     """
+    if not (_twopc_state is cp.TwoPCFsmStart) then
+      @printf[I32]("2PC: ERROR: _twopc_state = %d\n".cstring(), _twopc_state())
+      Fail()
+    end
+
+    ifdef "checkpoint_trace" then
+      @printf[I32]("2PC: Checkpoint state %s at ConnectorSink %s, txn-id %s\n".cstring(), checkpoint_id.string().cstring(), _sink_id.string().cstring(), _twopc_txn_id.cstring())
+    end
+
+    let wb: Writer = wb.create()
+    wb.u64_be(_twopc_current_offset.u64())
+    wb.u64_be(_notify.acked_point_of_ref)
+    let bs = wb.done()
+
     _event_log.checkpoint_state(_sink_id, checkpoint_id,
-      recover val Array[ByteSeq] end)
+      consume bs)
 
   be prepare_for_rollback() =>
-    _prepare_for_rollback()
+    ifdef "checkpoint_trace" then
+      @printf[I32]("Prepare for checkpoint rollback at ConnectorSink %s\n".cstring(), _sink_id.string().cstring())
+    end
+    //// _clear_barriers() will put a normal message processor
+    //// in place, and if a message arrives at process_message()
+    //// before we get the rollback() message, then we can send
+    //// bogus data that will be thrown away by the connector
+    //// and/or confused the connector sink and perhaps trigger
+    //// further 2PC phase 1 aborts.
 
   fun ref _prepare_for_rollback() =>
     _clear_barriers()
 
+  be incremental_rollback(payload: ByteSeq val, event_log: EventLog,
```

Comment:
A `Resilient` doesn't need to implement this behavior if it doesn't engage in incremental checkpointing.

Reply:
Oh, the default trait implementation also fails. I'll remove this.